src/Processors/Formats/Impl/JSONColumnsBaseBlockInputFormat.h

```diff
@@ -0,0 +1,92 @@
+namespace DB
+{
+
+class ReadBuffer;
+
+
+/// Base class for reading data in Columnar JSON formats.
+class JSONColumnsBaseReader
+{
+public:
+    JSONColumnsBaseReader(ReadBuffer & in_);
+
+    virtual ~JSONColumnsBaseReader() = default;
+
+    void setReadBuffer(ReadBuffer & in_) { in = &in_; }
+
+    virtual void readChunkStart() = 0;
+    virtual std::optional<String> readColumnStart() = 0;
+
+    virtual bool checkChunkEnd() = 0;
+    bool checkChunkEndOrSkipColumnDelimiter();
+
+    bool checkColumnEnd();
+    bool checkColumnEndOrSkipFieldDelimiter();
+
+    void skipColumn();
+
+protected:
+    ReadBuffer * in;
+};
+
+
+/// Base class for Columnar JSON input formats. It works with data using
+/// JSONColumnsBaseReader interface.
+/// To implement new Columnar JSON format you need to implement new JSONColumnsBaseReader
+/// interface and provide it to JSONColumnsBaseBlockInputFormat.
+class JSONColumnsBaseBlockInputFormat : public IInputFormat
+{
+public:
+    JSONColumnsBaseBlockInputFormat(ReadBuffer & in_, const Block & header_, const FormatSettings & format_settings_, std::unique_ptr<JSONColumnsBaseReader> reader_);
+
+    String getName() const override { return "JSONColumnsBaseBlockInputFormat"; }
+
+    void setReadBuffer(ReadBuffer & in_) override;
+
+    const BlockMissingValues & getMissingValues() const override { return block_missing_values; }
+
+protected:
+    Chunk generate() override;
+
+    size_t readColumn(IColumn & column, const DataTypePtr & type, const SerializationPtr & serialization, const String & column_name);
+
+    const FormatSettings format_settings;
+    const NamesAndTypes fields;
+    /// Maps column names and their positions in header.
+    std::unordered_map<String, size_t> name_to_index;
+    Serializations serializations;
+    std::unique_ptr<JSONColumnsBaseReader> reader;
+    BlockMissingValues block_missing_values;
+};
+
+
+/// Base class for schema inference from Columnar JSON input formats. It works with data using
+/// JSONColumnsBaseReader interface.
+/// To implement schema reader for the new Columnar JSON format you need to implement new JSONColumnsBaseReader
+/// interface and provide it to JSONColumnsBaseSchemaReader.
+class JSONColumnsBaseSchemaReader : public ISchemaReader
+{
+public:
+    JSONColumnsBaseSchemaReader(ReadBuffer & in_, const FormatSettings & format_settings_, std::unique_ptr<JSONColumnsBaseReader> reader_);
+
+private:
+    NamesAndTypesList readSchema() override;
+
+    /// Read whole column in the block (up to max_rows_to_read rows) and extract the data type.
+    DataTypePtr readColumnAndGetDataType(const String & column_name, size_t & rows_read, size_t max_rows_to_read);
+
+    /// Choose result type for column from two inferred types from different rows.
+    void chooseResulType(DataTypePtr & type, const DataTypePtr & new_type, const String & column_name, size_t row) const;
+
+    const FormatSettings format_settings;
+    std::unique_ptr<JSONColumnsBaseReader> reader;
```

Comment:
Hm, why an unique_ptr? A normal value member (maybe moved into the ctor in l. 77) should do the trick.

Reply:
Because JSONColumnsBaseReader is a base class, so we need a possibility to use it's subclasses here and unique_ptr is the best way to do it, we can just call JSONColumnsBaseSchemaReader constructor with unique_ptr that contains subclass of JSONColumnsBaseReader. It won't work with normal value member.